### PR TITLE
R sched

### PR DESCRIPTION
**Pull request by mfeurer**

### CODE REVIEW

File: smac/epm/gaussian_process_mcmc.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+from copy import deepcopy
+
+from smac.epm.base_gp import BaseModel
+from smac.epm.gaussian_process import GaussianProcess
+from smac.utils import normalization
+
+logger = logging.getLogger(__name__)
+
+
+class GaussianProcessMCMC(BaseModel):
```
Comment (KEggensperger) [inline smac/epm/gaussian_process_mcmc.py]: If we use both GP classes, this one should inherit from the other or both should be merged.

File: smac/epm/reinterpolation_rf.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
+import numpy as np
+import logging
+
+from pyrfr import regression
+
+from smac.epm.base_epm import AbstractEPM
+from smac.epm.rf_with_instances import RandomForestWithInstances
+from smac.utils.constants import N_TREES
+
+__author__ = "Aaron Klein"
```
Comment (KEggensperger) [inline smac/epm/reinterpolation_rf.py]: seems to be wrong

File: smac/initial_design/multi_config_initial_design.py [hunk 3]
```diff
@@ -60,14 +62,18 @@ def __init__(self,
...
             Intensification object to issue a racing to decide the current
             incumbent.
         aggregate_func: typing:Callable
             Function to aggregate performance of a configuration across
             instances.
+        configs: typing.Optional[typing.List[Configuration]]
+            List of initial configurations.
+        n_configs_x_params: int
+            how many configurations will be used at most in the initial design (X*D)
+        max_config_fracs: float
```
Comment (KEggensperger) [inline smac/initial_design/multi_config_initial_design.py]: mention that if we use wallclocklimit that this limit is obsolete

File: smac/optimizer/base_prior.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
+
+import numpy as np
+import scipy.stats as sps
+
+
+class BasePrior(object):
```
Comment (KEggensperger) [inline smac/optimizer/base_prior.py]: could we call this file gp_base_priors?

File: smac/optimizer/default_priors.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+
```
Comment (KEggensperger) [inline smac/optimizer/default_priors.py]: same here. Do we need all of them?

File: smac/optimizer/ei_optimization.py [hunk 6]
```diff
@@ -319,6 +317,73 @@ def _one_iter(
 
         return acq_val_incumbent, incumbent
 
+class DiffOpt(AcquisitionFunctionMaximizer):
```
Comment (KEggensperger) [inline smac/optimizer/ei_optimization.py]: Do we use this somewhere?

File: smac/optimizer/random_configuration_chooser.py [hunk 7]
```diff
@@ -64,3 +69,65 @@ def check(self, iteration: int):
             return True
         else:
             return False
+        
+class ChooserProb(RandomConfigurationChooser):
+
```
Comment (KEggensperger) [inline smac/optimizer/random_configuration_chooser.py]: These don't have docstrings

File: smac/utils/io/cmd_reader.py [hunk 8]
```diff
@@ -374,7 +374,7 @@ def _add_main_options(self):
                               default=logging.INFO, choices=["INFO", "DEBUG"],
                               help="Verbosity level.")
         opt_opts.add_argument("--mode",
-                              default="SMAC", choices=["SMAC", "ROAR", "EPILS", "Hydra", "PSMAC"],
+                              default="SMAC", choices=["SMAC", "ROAR", "EPILS", "Hydra", "PSMAC", "BORF"],
```
Comment (KEggensperger) [inline smac/utils/io/cmd_reader.py]: BOGP?

File: smac/utils/constants.py [hunk 9]
```diff
@@ -1,3 +1,5 @@
 MAXINT = 2 ** 31 - 1
 MINIMAL_COST_FOR_LOG = 0.00001 
-MAX_CUTOFF = 65535
\ No newline at end of file
+MAX_CUTOFF = 65535
```
Comment (KEggensperger) [inline smac/utils/constants.py]: Could we explain why this is 65535

File: smac/epm/base_gp.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
+import abc
```
[the remark this answers is not in the document]
Comment (mlindauer) [reply smac/epm/base_gp.py]: Yes, I would think that makes sense for the momemt. Both GP classes inherit from this base class.

File: smac/epm/gaussian_process_mcmc.py (hunk 1 above)
Comment (mlindauer) [reply smac/epm/gaussian_process_mcmc.py]: is there any code duplication if we don't merge both classes? it does not seem so.

File: smac/epm/gaussian_process_mcmc.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        # positive float value
+        if v.shape[0] == 1:
+            v = np.clip(v, np.finfo(v.dtype).eps, np.inf)
+        else:
+            v = np.clip(v, np.finfo(v.dtype).eps, np.inf)
+            v[np.where((v < np.finfo(v.dtype).eps) & (v > -np.finfo(v.dtype).eps))] = 0
+
+        return m, v
+    
+    def predict_marginalized_over_instances(self, X_test, **kwargs):
```
Comment (mlindauer) [inline smac/epm/gaussian_process_mcmc.py]: @KEggensperger @AndreBiedenkapp 
TODOs
1. move to base_gp
2. add docstring
3. add warning in GP facade saying that our GPs don't support instances

File: smac/facade/smac_facade.py [hunk 12]
```diff
@@ -402,27 +410,36 @@ def __init__(self,
...
+            Random number generator
 
         Returns
         -------
-        ChooserNoCoolDown|ChooserLinearCoolDown
+        RandomConfigurationChooser
 
         """
         if not random_configuration_chooser:
+            #return ChooserCosineAnnealing(prob_max=0.5, prob_min=0.001,
```
Comment (mfeurer) [inline smac/facade/smac_facade.py]: We should remove all commented out code.

File: smac/initial_design/factorial_design.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+__author__ = "Marius Lindauer"
+__copyright__ = "Copyright 2016, ML4AAD"
+__license__ = "3-clause BSD"
+
+
+class FactorialInitialDesign(MultiConfigInitialDesign):
+    """ Factorial initial design
+
+    Attributes
```
Comment (mfeurer) [inline smac/initial_design/factorial_design.py]: Shouldn't this be parameters?

File: smac/initial_design/factorial_design.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+            conf = deactivate_inactive_hyperparameters(conf_dict, cs)
+            conf.origin = "Factorial Design"
+            configs.append(conf)
+            self.logger.debug(conf)
+
+        self.logger.debug("Size of factorial design: %d" %(len(configs)))
+            
+        return configs
+            
+
```
Comment (mfeurer) [inline smac/initial_design/factorial_design.py]: Too many new lines.

File: smac/initial_design/latin_hypercube_design.py [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+
+__author__ = "Marius Lindauer"
+__copyright__ = "Copyright 2018, ML4AAD"
+__license__ = "3-clause BSD"
+
+
+class LHDesign(MultiConfigInitialDesign):
+    """ Latin Hypercube design
+
+    Attributes
```
Comment (mfeurer) [inline smac/initial_design/latin_hypercube_design.py]: Shouldn't this be parameters?

File: smac/initial_design/latin_hypercube_design.py [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+            conf = deactivate_inactive_hyperparameters(conf_dict, cs)
+            conf.origin = "LHD"
+            configs.append(conf)
+            self.logger.debug(conf)
+
+        self.logger.debug("Size of lhd: %d" %(len(configs)))
+            
+        return configs
+            
+
```
Comment (mfeurer) [inline smac/initial_design/latin_hypercube_design.py]: Too many newlines.

File: smac/initial_design/multi_config_initial_design.py [hunk 17]
```diff
@@ -38,9 +38,11 @@ def __init__(self,
                  traj_logger: TrajLogger,
                  runhistory: RunHistory,
                  rng: np.random.RandomState,
-                 configs: typing.List[Configuration],
                  intensifier: Intensifier,
-                 aggregate_func: typing.Callable
+                 aggregate_func: typing.Callable,
+                 configs: typing.Optional[typing.List[Configuration]]=None,
+                 n_configs_x_params: int = 10,
+                 max_config_fracs: float = 0.25
```
Comment (mfeurer) [inline smac/initial_design/multi_config_initial_design.py]: No whitespace around the assignment operator.

File: smac/initial_design/multi_rand_design.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    MultiConfigInitialDesign
+
+
+__author__ = "Marius Lindauer"
+__copyright__ = "Copyright 2018, ML4AAD"
+__license__ = "3-clause BSD"
+
+
+class MultiRandDesign(MultiConfigInitialDesign):
+    """ multi random design
```
Comment (mfeurer) [inline smac/initial_design/multi_rand_design.py]: How about a longer docstring?

File: smac/initial_design/multi_rand_design.py [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+    Attributes
```
Comment (mfeurer) [inline smac/initial_design/multi_rand_design.py]: Parameters?

File: smac/initial_design/sobol_design.py [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+__author__ = "Marius Lindauer"
+__copyright__ = "Copyright 2018, ML4AAD"
+__license__ = "3-clause BSD"
+
+
+class SobolDesign(MultiConfigInitialDesign):
+    """ Sobol sequence design
+
+    Attributes
```
Comment (mfeurer) [inline smac/initial_design/sobol_design.py]: Parameters?

File: smac/initial_design/sobol_design.py [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+            conf = deactivate_inactive_hyperparameters(conf_dict, cs)
+            conf.origin = "Sobol"
+            configs.append(conf)
+            self.logger.debug(conf)
+
+        self.logger.debug("Length of Sobol sequence: %d" %(len(configs)))
+            
+        return configs
+            
+
```
Comment (mfeurer) [inline smac/initial_design/sobol_design.py]: Too many newlines.

File: smac/optimizer/base_prior.py (hunk 4 above)
Comment (mfeurer) [reply smac/optimizer/base_prior.py]: And move this file to the epm directory?

Comment (mfeurer) [reply smac/optimizer/base_prior.py]: And also add type hints to all GP files?

File: smac/optimizer/default_priors.py (hunk 5 above)
Comment (mfeurer) [reply smac/optimizer/default_priors.py]: And also same for my previous suggestions.

File: smac/optimizer/random_configuration_chooser.py [hunk 22]
```diff
@@ -45,8 +48,10 @@ def check(self, iteration):
 class ChooserLinearCoolDown(RandomConfigurationChooser):
 
     def __init__(self, start_modulus: float=2.0, modulus_increment: float=0.3, end_modulus: float=float('inf')):
+        self.logger = logging.getLogger(
+            self.__module__ + "." + self.__class__.__name__)
         if start_modulus <= 1.0 and modulus_increment <= 0.0:
-            logging.warning("Using SMAC with random configurations only."
+            self.logger.warning("Using SMAC with random configurations only."
                             "ROAR is the better choice for this.")
```
Comment (mfeurer) [inline smac/optimizer/random_configuration_chooser.py]: Bad indentation.

File: smac/optimizer/random_configuration_chooser.py [hunk 23]
```diff
@@ -64,3 +69,65 @@ def check(self, iteration: int):
             return True
         else:
             return False
+        
+class ChooserProb(RandomConfigurationChooser):
```
Comment (mfeurer) [inline smac/optimizer/random_configuration_chooser.py]: There should be two newlines in front of a class definition.

File: smac/runhistory/runhistory2epm.py [hunk 24]
```diff
@@ -362,7 +362,158 @@ def _build_matrix(self, run_dict: typing.Mapping[RunKey, RunValue],
                 " log cost." % (constants.MINIMAL_COST_FOR_LOG))
             y[y < constants.MINIMAL_COST_FOR_LOG] =\
                 constants.MINIMAL_COST_FOR_LOG
-        y = np.log10(y)
+        y = np.log(y)
+
+        return X, y
+
+class RunHistory2EPM4ScaledCost(RunHistory2EPM4Cost):
+    """TODO"""
```
Comment (mfeurer) [inline smac/runhistory/runhistory2epm.py]: TODO!

File: smac/runhistory/runhistory2epm.py [hunk 25]
```diff
@@ -362,7 +362,158 @@ def _build_matrix(self, run_dict: typing.Mapping[RunKey, RunValue],
...
+            perc = np.percentile(y, 5)
+            min_y = 2 * np.min(y) - perc # ensure that scaled y cannot be 0
+            max_y = np.max(y)
+            # linear scaling
+            y = (y - min_y) / (max_y - min_y)
+
+        return X, y
+
+class RunHistory2EPM4InvScaledCost(RunHistory2EPM4Cost):
+    """TODO"""
```
Comment (mfeurer) [inline smac/runhistory/runhistory2epm.py]: TODO!

File: smac/runhistory/runhistory2epm.py [hunk 26]
```diff
@@ -362,7 +362,158 @@ def _build_matrix(self, run_dict: typing.Mapping[RunKey, RunValue],
...
+            min_y = 2 * np.min(y) - perc # ensure that scaled y cannot be 0
+            max_y = np.max(y)
+            # linear scaling
+            y = (y - min_y) / (max_y - min_y)
+            y = 1 - 1/y
+
+        return X, y
+    
+class RunHistory2EPM4SqrtScaledCost(RunHistory2EPM4Cost):
+    """TODO"""
```
Comment (mfeurer) [inline smac/runhistory/runhistory2epm.py]: TODO!

File: smac/smac_cli.py [hunk 27]
```diff
@@ -115,6 +115,15 @@ def main_cli(self, commandline_arguments: typing.List[str]=None):
                 stats=stats,
                 restore_incumbent=incumbent,
                 run_id=main_args_.seed)
+        if main_args_.mode == "BORF":
```
Comment (mfeurer) [inline smac/smac_cli.py]: Also BOGP?

File: smac/utils/normalization.py [hunk 28]
```diff
@@ -0,0 +1,32 @@
+import numpy as np
```
Comment (mfeurer) [inline smac/utils/normalization.py]: Please move this file to emp.py.

File: smac/epm/gaussian_process_mcmc.py (hunk 11 above)
Comment (KEggensperger) [reply smac/epm/gaussian_process_mcmc.py]: Done with 6a16ececf5dc1d3101160019c907e8103de9332f

File: smac/facade/bogp_facade.py [hunk 29]
```diff
@@ -0,0 +1,187 @@
...
+
+    """
+
+    def __init__(self, model_type='gp_mcmc', **kwargs):
+        """
+        Constructor
+        see ~smac.facade.smac_facade for docu
+        """
+        if 'model' not in kwargs or kwargs['model'] is None:
+            self.rng = np.random.RandomState(np.random.randint(0, 10000))
```
[the remark this answers is not in the document]
Comment (KEggensperger) [reply smac/facade/bogp_facade.py]: Done in b9ddd52

**Pull request merged**